cognite/client/_api/datapoints.py

```diff
@@ -856,3 +890,125 @@ def _get_datapoints(
         expected_fields = [a for a in aggs] if aggs is not None else ["value"]
         dps = Datapoints._load(res, expected_fields, cognite_client=self.client._cognite_client)
         return dps
+
+
+class _DFColQuery:
+    def __init__(self, start, end, ts_item, aggregates, granularity):
+        self.start = cognite.client.utils._time.timestamp_to_ms(start)
+        self.end = cognite.client.utils._time.timestamp_to_ms(end)
+        self.ts_item = ts_item
+        self.aggregates = aggregates
+        self.granularity = granularity
+        self.granularity_ms = cognite.client.utils._time.granularity_to_ms(granularity)
+        self.result = None
+        self.exception = None
+
+
+class DataFrameFetcher:
+    START_JOBS_PER_THREAD = 2  # how many jobs per worker thread to aim for at the start
+    SLEEP_WAIT_NEW_JOBS = 0.010  # how long to sleep while waiting for new jobs
+
+    def __init__(self, client):
+        self.np, self.pd = utils._auxiliary.local_import("numpy", "pandas")
+        self.client = client
+        self.jobs = []
```

Comment:
I would prefer if self.jobs was a Queue, and you add jobs to the queue as you progress. That way you don't have to worry about locking and marking workers as idle.

Reply:
Queue vs stack doesn't matter much here, but if we keep this, but the idle has a use, let's discuss.
Replacing the idle logic with something more proper (semaphore?) is an idea, but something for later I think

Reply:
see what you mean now regarding queue, using that now and got rid of the job queue-in-a-job queue logic